[5.7-04182022] Layout distributed actor id and actorSystem fields first in stored properties

DougGregor opened the pull request:
The stored properties `id` and `actorSystem` that are synthesized for
a distributed actor would end up getting placed in fairly random
locations within the stored properties lists based on when they were
synthesized. This means that they (1) weren't always in the prefix, and
(2) weren't always the same order in every translation unit. Hilarity ensues.

Original PR: https://github.com/apple/swift/pull/42567

Fixes rdar://92142457.

DougGregor committed:
Layout distributed actor id and actorSystem fields first in stored properties

The stored properties `id` and `actorSystem` that are synthesized for
a distributed actor would end up getting placed in fairly random
locations within the stored properties lists based on when they were
synthesized. This means that they (1) weren't always in the prefix, and
(2) weren't always the same order in every translation unit. Hilarity ensues.

Fixes rdar://92142457.

DougGregor commented:
@swift-ci please test

ktoso commented:
```

Failed Tests (5):
  Swift(macosx-x86_64) :: SourceKit/Sema/sema_regex.swift
  Swift(macosx-x86_64) :: StringProcessing/Frontend/enable-flag.swift
  Swift(macosx-x86_64) :: StringProcessing/Parse/forward-slash-regex.swift
  Swift(macosx-x86_64) :: StringProcessing/Runtime/regex_basic.swift
  Swift(macosx-x86_64) :: StringProcessing/Sema/regex_literal_type_inference.swift
```

ktoso commented:
@swift-ci please test macOS

ktoso committed:
Always emit id and actorSystem in expected order

Without this change, the results of the `enumerateStoredPropertiesAndMissing` function still might return in the wrong order.

Specifically, previously we would emit the stored, synthesized properties in the right order:

```
(var_decl implicit "id" type='SampleSystem.ActorID' interface type='SampleSystem.ActorID' access=public let readImpl=stored immutable
  (accessor_decl implicit 'anonname=0x1372fc578' interface type='(Worker_OTHER_MODULE) -> () -> SampleSystem.ActorID' ...

(var_decl implicit "actorSystem" type='Worker_OTHER_MODULE.ActorSystem' interface type='Worker_OTHER_MODULE.ActorSystem' access=public let readImpl=stored immutable
  (accessor_decl implicit 'anonname=0x1372fc908' interface type='(Worker_OTHER_MODULE) -> () -> Worker_OTHER_MODULE.ActorSystem' ...
```

However, the results would still be in the "normal" order if the properties were from the interface (I think because the protocol conformance?), rather than the concrete nominal... This would manifest in results looking like this:

```
(var_decl implicit "actorSystem" type='Worker_OTHER_MODULE.ActorSystem' interface type='Worker_OTHER_MODULE.ActorSystem' access=public let readImpl=stored immutable
  (accessor_decl implicit 'anonname=0x147186770' access=public final get_for=actorSystem
    (parameter "self")
    (parameter_list)))
(var_decl implicit "id" type='SampleSystem.ActorID' interface type='SampleSystem.ActorID' access=public let readImpl=stored immutable
  (accessor_decl implicit 'anonname=0x14718aa18' access=public final get_for=id
    (parameter "self")
    (parameter_list)))
```

Since those are not implicit, and just getters for the properties they would emitted in arbitrary order again. While this should not matter I guess... it definitely does, as fixing the order of this causes `irgen::forEachField` to emit them in the right order always.

I have confirmed this in a real project where the types in IRGen now align and are consistently correct:

```
-> % cat main.ir | grep Worker | grep '= type <'
%T3Lib19Worker_OTHER_MODULEC = type <{ %swift.refcounted, %swift.defaultactor, %T3Lib20THE_ACTOR_IDENTIFIERV, %T3Lib12SampleSystemC* }>

-> % cat othermodule.ir | grep Worker | grep '= type <'
%T3Lib19Worker_OTHER_MODULEC = type <{ %swift.refcounted, %swift.defaultactor, %T3Lib20THE_ACTOR_IDENTIFIERV, %T3Lib12SampleSystemC* }>
```

I am not very sure how to write a test covering this but will try to do this next as we would not want to regress this under any circumstances.

ktoso commented:
Added a minor change with full explanation about it in the commit message: https://github.com/apple/swift/commit/f58a64f92a3e7adce9e1459bd5c7e3fc8dcfb9ab

With this patch I have confirmed in a real project this solves the cross module issue

hamishknight commented:
@swift-ci please test

ktoso commented:
@swift-ci please build toolchain macOS
